Introduce OpenTelemetry observability service (#2)
dynatrace-oss-contrib/sdk-go: observability/opentelemetry/v2/README.md

# OpenTelemetry instrumentation for CloudEvents

This package contains the components necessary to instrument CloudEvents clients with OpenTelemetry. The main component is the `OTelObservabilityService` which implements the `ObservabilityService` interface from CloudEvents.

> ## Getting a fully instrumented CloudEvents HTTP client

Reviewer: ```suggestion
## Instrumented CloudEvents HTTP client
```

# OpenTelemetry instrumentation for cloudevents

This package contains the components necessary to instrument cloudevents clients with OpenTelemetry. The main component is the `OTelObservabilityService` which implements the `ObservabilityService` interface from cloudevents.

## Getting a fully instrumented cloudevents HTTP client

If you want to get a fully instrumented HTTP client, use the helper method in the ` github.com/cloudevents/sdk-go/observability/opentelemetry/v2` module:

```go
import (
	"context"

	otelObs "github.com/cloudevents/sdk-go/observability/opentelemetry/v2/client"
	cloudevents "github.com/cloudevents/sdk-go/v2"
	"github.com/cloudevents/sdk-go/v2/client"
	cehttp "github.com/cloudevents/sdk-go/v2/protocol/http"
)

// you can pass the http/client options as usual
c, _ := otelObs.NewClientHTTP([]cehttp.Option{}, []client.Option{})
```

This will produce spans for all outgoing and incoming requests. By default, the spans will have the attributes as defined in [keys.go](https://github.com/cloudevents/sdk-go/blob/release-2.5/v2/observability/keys.go). For more advanced configuration, see the next section.

## Advanced configuration

### HTTP auto-instrumentation

> In order to generate spans when sending and receiving events, it's necessary to configure the HTTP client from cloudevents with OpenTelemetry instrumentation. The client has two "possibilities" for instrumentation:

Reviewer: ```suggestion
In order to generate spans when sending and receiving events, it's necessary to configure the HTTP client from cloudevents with OpenTelemetry instrumentation. The client has two potentially interesting points for instrumentation:
```
maybe?

```go
p, err := cloudevents.NewHTTP(
	cloudevents.WithRoundTripper(otelhttp.NewTransport(http.DefaultTransport)),
	cloudevents.WithMiddleware(func(next http.Handler) http.Handler {
		return otelhttp.NewHandler(next, "receive")
	}),
)
```

The `otelhttp.NewTransport` will ensure that spans are generated for each outgoing request, and that the `traceparent` header is properly propagated. The `otelhttp.NewHandler` will take care of incoming requests, reading the `traceparent` header and continuing the trace with a new span.

This already gives some observability "out-of-the-box", but the spans generated only contain common HTTP headers as defined in the [HTTP semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.6.1/specification/trace/semantic_conventions/http.md). Also if you are using another protocol, then spans will not be automatically generated, unless there is an auto-instrumentation library for that.

Because of this, cloudevents offers the `ObservabilityService` interface which is used to generate spans, independently of the chosen protocol. See next how to configure the cloudevents client to use it.

> 

Reviewer: ```suggestion
```

### Using the OTelObservabilityService

The most basic way to configure the cloudevents client to use the `OTelObservabilityService` is:

```go
c, err := cloudevents.NewClient(p, client.WithObservabilityService(otelObs.NewOTelObservabilityService()))
```

With the above configuration, the spans generated by the `OTelObservabilityService` will have:

<table>
  <tbody>
    <tr>
      <td style="font-weight:bold">Span name</td>
      <td><code>cloudevents.client.[event type] [operation name]</code> where <code>[operation name]</code> follows the OpenTelemetry <a href="https://github.com/open-telemetry/opentelemetry-specification/blob/v1.6.1/specification/trace/semantic_conventions/messaging.md#operation-names">semantic conventions for messaging systems</a></td>
    </tr>
    <tr>
      <td style="font-weight:bold">Span attributes</td>
      <td>The attributes as defined in <a href="https://github.com/cloudevents/sdk-go/blob/release-2.5/v2/observability/keys.go">keys.go</a></td>
    </tr>

>     </tbody>

Reviewer: ```suggestion
  </tbody>
```

</table>

If you require different span names or extra attributes, you can pass multiple `OTelObservabilityServiceOption` options when creating the observability service:

```go
nameFormatter := func(e *cloudevents.Event) string {
	return "my.custom.name." + e.Context.GetType()
}

attributesGetter := func(*cloudevents.Event) []attribute.KeyValue {
	return []attribute.KeyValue{
		attribute.String("my-attr", "some-value"),
	}
}

// create the obs service with custom span names and attributes
os := otelObs.NewOTelObservabilityService(
	otelObs.WithSpanNameFormatter(nameFormatter),
	otelObs.WithSpanAttributesGetter(attributesGetter),
)

c, err := cloudevents.NewClient(p, client.WithObservabilityService(os))
```

> >NOTE: The `nameFormatter` and `attributesGetter` functions will be called on each span creation. **Avoid** doing any heavy processing in them.

Reviewer: ```suggestion
>Note: The `nameFormatter` and `attributesGetter` functions will be called on each span creation. **Avoid** doing any heavy processing in them.
```
Or use `NOTE` in the two occurrences down below, so the casing is consistent.

> 

Reviewer: ```suggestion
```

## Extra types

This package also contains extra types and helper functions that are useful in case you need to access/set the `tracecontext` in a more "low-level" way.

> They allow to inject(write) and extract(read) `tracecontext` from the event. This is particularly useful when dealing with code that has no notion of a "request" nor a context. For example, long-running background processes pooling from a queue. 

Reviewer: ```suggestion
They allow to inject(write) and extract(read) `tracecontext` from the event. This is particularly useful when dealing with code that has no notion of a "request" nor a context. For example, long-running background processes polling from a queue. 
```
I might be wrong about which verb you wanted to use here, so please ignore if this is not what you meant :sweat_smile:

Author: No, you are correct 😁

>Note: To learn more about the propagation, take a look at the [Propagators API SPEC](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.6.1/specification/context/api-propagators.md).

> 

Reviewer: ```suggestion
```

## Manually extracting/injecting tracecontext from the event

When working with distributed systems, it can be difficult to achieve proper context propagation. For example, a long-running process listening to a topic does not have a "context" concept like a HTTP server receiving requests does.

> >Note: The OpenTelemetry community is always creating auto-instrumentation libraries for popular SDKs and frameworks, so there could be one for the queue client or other framework you are using. It's worth checking out what is available in the [go-contrib GitHub repository](https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/instrumentation)

Reviewer: Would it also make sense to link to [the OpenTelemetry Registry](https://opentelemetry.io/registry/) here? E.g.: the Queue system has an Auto-instrumentation available out of the box, or would that not work with cloudevents?

Author: AFAIK, the way cloudevents works (at least this one for go) is the `client.go` they offer has a transport that can be many things. HTTP, amqp, nats, kafka, etc. Since there's many protocols, I believe the idea with this Observability service is to always have spans being created when using the client, since it's the same for all protocols, serving as as central point.

I guess listing the Registry also could work. Since this is for Go, I thought of just putting the contrib link but maybe something else goes into the "core" one so it gets missed.

I added auto-instrumentation for HTTP, so incoming/outgoing requests have propagation working. If there are libraries for the other protocols, those could also be added later on for sure.

Author: I rephrased it a bit stating the registry and as support increases, new auto-instrumentation libraries can be added.

It can be used to get access to the "raw" `tracecontext` values (`traceparent` and `tracestate`). One use case is to inject the `tracecontext` from the `context` into the carrier, to gain access to the populated `DistributedTracingExtension`:

```go
type MyEvent struct { 
	TraceParent string `json:"traceparent,omitempty"`
	TraceState string `json:"tracestate,omitempty"`
}

func injectAndReadTraceParentAndState(ctx context.Context, e cloudevents.Event) {

    me := MyEvent{}

	// the propagator from OpenTelemetry
    prop := propagation.TraceContext{}

	carrier := otelObs.NewCloudEventCarrier()

    // Injects (writes) the tracecontext into the NewCloudEventCarrier
    // Doing so, will set the DistributedTracingExtension fields
	prop.Inject(ctx, carrier)

    // Here then we have the "raw" access to the tracecontext data
    // https://www.w3.org/TR/trace-context/

    // e.g. 00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01
	me.TraceParent = carrier.Extension.TraceParent

    // e.g. congo=t61rcWkgMzE

> 	me.TraceState = carrier.Extension.TraceState

Reviewer: ```suggestion
type MyEvent struct {
	TraceParent string `json:"traceparent,omitempty"`
	TraceState  string `json:"tracestate,omitempty"`
}

func injectAndReadTraceParentAndState(ctx context.Context, e cloudevents.Event) {

	me := MyEvent{}

	// the propagator from OpenTelemetry
	prop := propagation.TraceContext{}

	carrier := otelObs.NewCloudEventCarrier()

	// Injects (writes) the tracecontext into the NewCloudEventCarrier
	// Doing so, will set the DistributedTracingExtension fields
	prop.Inject(ctx, carrier)

	// Here then we have the "raw" access to the tracecontext data
	// https://www.w3.org/TR/trace-context/

	// e.g. 00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01
	me.TraceParent = carrier.Extension.TraceParent

	// e.g. congo=t61rcWkgMzE
	me.TraceState = carrier.Extension.TraceState
}
```

Consistent tab/spaces formatting

Author: Fixed.
